runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/LazyJobBundleFactory.java

```diff
@@ -53,92 +55,81 @@
+  // Using environment as the initialization marker.
+  private Environment environment;
+  private ExecutorService executor;
+  private GrpcFnServer<FnApiControlClientPoolService> controlServer;
+  private GrpcFnServer<GrpcLoggingService> loggingServer;
+  private GrpcFnServer<ArtifactRetrievalService> retrievalServer;
+  private GrpcFnServer<StaticGrpcProvisionService> provisioningServer;
+  private GrpcFnServer<GrpcDataService> dataServer;
+  private GrpcFnServer<GrpcStateService> stateServer;
+  private MapControlClientPool clientPool;
+  private EnvironmentFactory environmentFactory;
 
-  JobBundleFactoryBase(JobInfo jobInfo) throws Exception {
-    ServerFactory serverFactory = getServerFactory();
-    IdGenerator stageIdGenerator = IdGenerators.incrementingLongs();
-    ControlClientPool clientPool = MapControlClientPool.create();
+  public static LazyJobBundleFactory create(
+      JobInfo jobInfo, Map<String, EnvironmentFactory.Provider> environmentFactoryProviderMap) {
+    return new LazyJobBundleFactory(jobInfo, environmentFactoryProviderMap);
+  }
 
-    GrpcFnServer<FnApiControlClientPoolService> controlServer =
-        GrpcFnServer.allocatePortAndCreateFor(
-            FnApiControlClientPoolService.offeringClientsToPool(
-                clientPool.getSink(), GrpcContextHeaderAccessorProvider.getHeaderAccessor()),
-            serverFactory);
-    GrpcFnServer<GrpcLoggingService> loggingServer =
-        GrpcFnServer.allocatePortAndCreateFor(
-            GrpcLoggingService.forWriter(Slf4jLogWriter.getDefault()), serverFactory);
-    GrpcFnServer<ArtifactRetrievalService> retrievalServer =
-        GrpcFnServer.allocatePortAndCreateFor(
-            BeamFileSystemArtifactRetrievalService.create(), serverFactory);
-    GrpcFnServer<StaticGrpcProvisionService> provisioningServer =
-        GrpcFnServer.allocatePortAndCreateFor(
-            StaticGrpcProvisionService.create(jobInfo.toProvisionInfo()), serverFactory);
-    EnvironmentFactory environmentFactory =
-        getEnvironmentFactory(
-            controlServer,
-            loggingServer,
-            retrievalServer,
-            provisioningServer,
-            clientPool.getSource(),
-            IdGenerators.incrementingLongs());
+  LazyJobBundleFactory(
+      JobInfo jobInfo, Map<String, EnvironmentFactory.Provider> environmentFactoryMap) {
+    IdGenerator stageIdGenerator = IdGenerators.incrementingLongs();
     this.stageIdGenerator = stageIdGenerator;
-    this.controlServer = controlServer;
-    this.loggingServer = loggingServer;
-    this.retrievalServer = retrievalServer;
-    this.provisioningServer = provisioningServer;
-    this.environmentCache = createEnvironmentCache(environmentFactory, serverFactory);
+    this.environmentCache =
+        createEnvironmentCache(
+            (environment) -> {
+              synchronized (this) {
+                checkAndInitialize(jobInfo, environmentFactoryMap, environment);
```

Comment:
I thought about it but avoided it because of
* This will result in interpretation of python/go pipeline options in flink-java. As the options are not standardized across the languages, this approach seems to be very fragile.
* Pipeline options are not always populated, especially in case of default options.
* This will make future support of mixed environment difficult in future.